lnwallet/channel_test.go

```diff
@@ -5801,6 +5801,58 @@ func TestChanReserve(t *testing.T) {
 	)
 }
 
+// TestChanReserveRemoteInitiator tests that the channel reserve of the
+// initiator is accounted for when adding HTLCs, whether the initiator is the
+// local or remote node.
+func TestChanReserveRemoteInitiator(t *testing.T) {
+	t.Parallel()
+
+	// We start out with a channel where both parties have 5 BTC.
+	aliceChannel, bobChannel, cleanUp, err := CreateTestChannels(
+		true,
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer cleanUp()
+
+	// Set Alice's channel reserve to be 5 BTC-commitfee. This means she
+	// has just enough balance to cover the comitment fee, but not enough
+	// to add any more HTLCs to the commitment. Although a reserve this
+	// high is unrealistic, a channel can easiliy get into a situation
+	// where the initiator cannot pay for the fee of any more HTLCs.
+	commitFee := aliceChannel.channelState.LocalCommitment.CommitFee
+	aliceMinReserve := 5*btcutil.SatoshiPerBitcoin - commitFee
+
+	aliceChannel.channelState.LocalChanCfg.ChanReserve = aliceMinReserve
+	bobChannel.channelState.RemoteChanCfg.ChanReserve = aliceMinReserve
+
+	// Now let Bob attempt to add an HTLC of 0.1 BTC. He has plenty of
+	// money, but Alice, which is the initiator, cannot afford any more
+	// HTLCs on the commitment transaction.
+	htlcAmt := lnwire.NewMSatFromSatoshis(0.1 * btcutil.SatoshiPerBitcoin)
+	htlc, _ := createHTLC(0, htlcAmt)
+
+	// Bob should refuse to add this HTLC, since he realizes it will create
+	// an invalid commitment.
+	_, err = bobChannel.AddHTLC(htlc, nil)
+	if err != ErrBelowChanReserve {
+		t.Fatalf("expected ErrBelowChanReserve, instead received: %v", err)
+	}
+
+	// Same for Alice, she should refuse to accept a state transtion with
+	// this second HTLC.
+	if _, err := aliceChannel.ReceiveHTLC(htlc); err != nil {
```

Comment:
Perhaps we should also check here when Alice attempts to add the HTLC? An earlier check means the link can be notified of this failure condition, and cancel back the incoming HTLC.

Reply:
You mean canceling the HTLC  immediately instead of waiting for the state transition? Not sure what we would gain from that, IIUC if Bob sends such an HTLC that is a protocol violation so we should fail the channel regardless. 

Would be a simple addition though, just calling `validateCommitmentSanity` in `ReceiveHTLC` I believe. let me know what you think would be best.

Reply:
Done: https://github.com/lightningnetwork/lnd/pull/3691/commits/1617dc769c17f7f489aa7fdd5fe408caccb45b8e